### PR TITLE
[wip] Master: Add support for per-node IP Pool utilization metric

### PR DESCRIPTION
**Pull request by abhat**
This commit adds support for per node IP pool utilization metric.

Signed-off by: Aniket Bhat <anbhat@redhat.com>

### CODE REVIEW

File: go-controller/pkg/ovn/pods.go [hunk 1]
```diff
@@ -381,5 +386,46 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
...
+// addLogicalSwitchPodCount updates the per node metric for ip utilization.
+// It also increments the podCount cache for the switch given by nodeName.
+func (oc *Controller) addLogicalSwitchPodCount(lsInfo *logicalSwitchInfo, nodeName string) error {
+	if lsInfo == nil {
+		return fmt.Errorf("Invalid logical switch info while adding pod count for node %s", nodeName)
+	}
+	if nodeName == "" {
+		return fmt.Errorf("Invalid node name provided while incrementing pod count")
+	}
+	lsInfo.podCount = lsInfo.podCount + 1
```
Comment (girishmg) [inline go-controller/pkg/ovn/pods.go]: This function is called exactly from one place and from where it is called `lsInfo` and `nodename` already have right values. So, you can remove lines 400-405 and this function shouldn't return `error`

File: go-controller/pkg/ovn/pods.go [hunk 2]
```diff
@@ -381,5 +386,46 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
...
+	oc.logicalSwitchCache[nodeName] = *lsInfo
+	return nil
+}
+
+// deleteLogicalSwitchPodCount updates the per node metric for ip utilization.
+// It also decrements the podCount cache for the switch given by nodeName by one.
+func (oc *Controller) deleteLogicalSwitchPodCount(nodeName string) error {
+	if nodeName == "" {
+		return fmt.Errorf("Invalid node name provided while decrementing pod count")
+	}
```
Comment (girishmg) [inline go-controller/pkg/ovn/pods.go]: nodeName will not be `empty` here, so perhaps you don't need the check above.

Also, I don't think this function should return an error. It should just log Warning and return.

**Comment by abhat** [comment]
> @abhat @dcbw The code makes an assumption that the `hostsubnet length` is same across various cluster subnet values. However, we don't enforce it. It could be that the cluster_subnet_range is 192.168.0.0/16/24 and 192.169.0.0/16/25. So, when this is the case `OvnClusterSubnetUsableIPs` is invalid as it will be set to 256-3 and on the other cluster subnet it should be 128-3.

That makes sense to me. I didn't think about the multiple ClusterCIDR case. I guess moving the host subnet length per node inside the `logicalSwitchInfo` struct should address that issue.

**Comment by girishmg** [comment]
@abhat or we should only allow same hostsubnet length across all the values of cluster subnet. I think this is the right thing to do. @dcbw what do you think?

**Comment by abhat** [comment]
> @abhat or we should only allow same hostsubnet length across all the values of cluster subnet. I think this is the right thing to do. @dcbw what do you think?

@dcbw any thoughts on this?

**Comment by dcbw** [comment]
@girishmg why would we need to enforce the same hostsubnet length for all subnets? That seems pretty restrictive, like if you install the cluster with a /16/24 and now you have 250 nodes, you have to find *another* /16/24 somewhere?

**Comment by girishmg** [comment]
> @girishmg why would we need to enforce the same hostsubnet length for all subnets? That seems pretty restrictive, like if you install the cluster with a /16/24 and now you have 250 nodes, you have to find _another_ /16/24 somewhere?

Not an _another_ /16/24, but maybe /23/24 if they need two additional nodes, or /22/24 if they need four additional nodes, and so on.

As a deployer, I already know the maximum number of pods I will support per node. By default, kubelet allows scheduling around 110 pods. If that default value is sufficient, then I would use /16/25. If I all need is 10 pods per node, then I will select /16/28. There could be other nodes that support even lesser pods, but the maximum number doesn't change.

The problem is that we don't have a way to pin a subnet to a node. A node will get the next available subnet from the provided cluster cidr ranges. If the host subnet length is same across cluster CIDRs, then we know that all the nodes are equal and can support `pow(2, 32-hostsunbetlength) -2` Pods.

Now, if we support having a different host subnet for cluster CIDR, then the nodes will end up supporting different number of pods. Since we don't have a way to pin a subnet to a node, we will not know which node in the k8s cluster got the tiny subnet (as compared to other nodes) and the kube scheduler will not know about the shortage in IPs when it schedules the pod.

**Comment by abhat** [comment]
Do we need to worry about making hard guarantees that the hostSubnetLength be the same across nodes? Seems like a constraint that is not really necessary unless it solves some design shortcoming? For purposes of this PR, I can very easily add the hostSubnetLength of each node in it's `logicalSwitchInfo` struct.

**Comment by trozet** [comment]
@abhat this has been around for a long time. Do you plan on updating this soon or can we close it?

**Comment by abhat** [comment]
It needs update. I will try to get to this next week.

**Comment by abhat** [comment]
We can close the original PR.

**Pull request closed without merging**